src/schemapack/_internals/spec/datapack.py

```diff
@@ -191,3 +188,42 @@ class DataPack(_FrozenNoExtraBaseModel):
             + " dependencies) of the root resource."
         ),
     )
+
+    @model_validator(mode="before")
+    @classmethod
+    def check_root_duality(cls, value: Mapping) -> Mapping | None:
+        """Ensures that both 'rootClass' and 'rootResource' are either present or absent."""
+        missing = [key for key in ("rootClass", "rootResource") if not value.get(key)]
+
+        if len(missing) == 1:
+            raise PydanticCustomError(
+                "DatapackRootDualityError",
+                "Invalid DataPack due to the missing field of '{missing}'.",
```

Comment:
```suggestion
                "Invalid DataPack due to missing field '{missing}'.",
```
This is a bit unspecific either way. Something similar to the doctring might be better.